aptos-core/consensus/src/block_storage/sync_manager.rs

```diff
@@ -785,27 +841,83 @@ impl BlockRetriever {
                     CUR_BLOCK_SYNC_BLOCK_SUM_GAUGE.with_label_values(&[]).add(batch.len() as i64);
                     result_blocks.extend(batch);
                     ledger_infos.extend(result.ledger_infos().clone());
+                    quorum_certs.extend(result.quorum_certs().clone());
                     break;
-                },
+                }
                 _e => {
                     bail!(
                         "Failed to fetch block {}, for original start {}",
                         last_block_id,
                         block_id,
                     );
-                },
+                }
             }
         }
-        if target_block_id != *GENESIS_BLOCK_ID {
-            assert_eq!(
-                result_blocks
-                    .last()
-                    .expect("Expected at least a result_block")
-                    .id(),
-                target_block_id
-            );
+        Ok((result_blocks, quorum_certs, ledger_infos))
+    }
+
+    async fn retrieve_block_by_epoch(
+        &mut self,
+        epoch: u64,
+        target_block_id: HashValue,
+        peers: Vec<AccountAddress>,
+        payload_manager: Arc<dyn TPayloadManager>,
+    ) -> anyhow::Result<(Vec<Block>, Vec<QuorumCert>, Vec<LedgerInfoWithSignatures>)> {
+        let mut result_blocks = vec![];
+        let mut ledger_infos = vec![];
+        let mut quorum_certs = vec![];
+        let response = self
+            .retrieve_block_for_id_chunk(
+                HashValue::zero(),
+                target_block_id,
+                self.max_blocks_to_request,
+                peers.clone(),
+                Some(epoch),
+            )
+            .await;
+        match response {
+            Ok(result) if matches!(result.status(), BlockRetrievalStatus::Succeeded) => {
```

Comment:
Would it retrieve all the blocks included in one epoch? if so, would it be huge memory consuming?